product-mini/platforms/posix/main.c

```diff
@@ -282,113 +282,124 @@ validate_env_str(char *env)
 #endif
 
 #if BH_HAS_DLFCN
-typedef uint32 (*get_native_lib_func)(char **p_module_name,
-                                      NativeSymbol **p_native_symbols);
-typedef int (*init_native_lib_func)(void);
-typedef void (*deinit_native_lib_func)(void);
+struct native_lib {
+    void *handle;
+
+    uint32 (*get_native_lib)(char **p_module_name,
+                             NativeSymbol **p_native_symbols);
+    int (*init_native_lib)(void);
+    void (*deinit_native_lib)(void);
 
-static uint32
-load_and_register_native_libs(const char **native_lib_list,
-                              uint32 native_lib_count,
-                              void **native_handle_list)
-{
-    uint32 i, native_handle_count = 0, n_native_symbols;
-    NativeSymbol *native_symbols;
     char *module_name;
-    void *handle;
+    NativeSymbol *native_symbols;
+    uint32 n_native_symbols;
+};
 
-    for (i = 0; i < native_lib_count; i++) {
-        /* open the native library */
-        if (!(handle = dlopen(native_lib_list[i], RTLD_NOW | RTLD_GLOBAL))
-            && !(handle = dlopen(native_lib_list[i], RTLD_LAZY))) {
-            LOG_WARNING("warning: failed to load native library %s",
-                        native_lib_list[i]);
-            continue;
-        }
+struct native_lib *
+load_native_lib(const char *name)
+{
+    struct native_lib *lib = malloc(sizeof(*lib));
```

Comment:
How about using wasm_runtime_malloc/wasm_runtime_free to be consistent with other places?